### DISCUSSION TITLE
Giving visibility into whether or not a package requires 2FA

### DISCUSSION THREAD

**Discussion by ianling** [npm/feedback#588] [Registry] [answered]
npm currently allows you to require 2FA when publishing a package or modifying a package's settings. https://docs.npmjs.com/requiring-2fa-for-package-publishing-and-settings-modification

This is essential for preventing account takeovers and package hijacking, like we saw today with the widely used package ua-parser-js: https://github.com/faisalman/ua-parser-js/issues/536

As a developer, it is impossible for me to have known whether the maintainer of ua-parser-js was following best practices in regards to securing their account, and by extension, securing the ua-parser-js package. In other words, I don't know if they had 2FA enabled.

This is a critical flaw that prevents me from being able to make an informed decision about whether or not I should introduce this dependency in my project. The inability to make informed decisions about code dependencies is one of the key problems that leads to supply chain attacks.

The only argument against this that I can think of is that it might lead to users being targeted due to their account's/package's lack of 2FA, however, I think this change would lead to a net increase in security through the power of peer pressure and public shaming.

Alternatively, starting on some arbitrary date, like 2022/01/01, all packages could be forced to have 2FA enabled in order to push changes.

Either way, what we have now clearly isn't a good solution for promoting good security posture and critical thinking about supply chain vulnerabilities.

**Comment by MylesBorins** [comment]
@ianling you are dead on with both the importance of this work and the privacy / security concerns with leaking details such as which projects have 2FA enabled.

We are actively discussing ways for us to improve 2FA adoption and overall security of the registry, but for the reasons you outlined above I do not believe that showing this status on a packages page is the correct approach.

Thanks for the suggestion!

**Reply by ianling** [reply]
Thanks for the response, glad to hear there are ongoing discussions around this stuff.

**Reply by kelset** [reply]
> for the reasons you outlined above I do not believe that showing this status on a packages page is the correct approach.

would maybe be better to add this "report" to the `npm audit` command? I think that it'd be important to have a way to quickly learn for a certain project which ones are not 2FA "secured"

**Reply by MylesBorins** [reply]
Whether it is via a UI or via audit, in either case this is information we should not expose publicly as it would make projects that do not have this bit a target.

**Reply by asklar** [reply]
I get the value of not exposing this information, but it's a little "security by obscurity". The packages that don't have 2FA are already susceptible. As a developer, I want to know whether anything I depend on lacks 2FA as this increases _my_ risk, so that I can find alternative packages. I think we're all in agreement that we should encourage increasing the security of the ecosystem, but I disagree with the assessment that showing which packages don't have 2FA will make the ecosystem less secure than it already is.

**Reply by MylesBorins** [reply]
@asklar I plan to bring this up with some folks internally from the security team to check my gut reaction on this. The intention is not "security through obscurity" but rather avoiding the disclosure of information that could make folks a target.

I wonder if there are other "signals" we can give regarding how packages are maintained / published that could give a similar feeling of good will without making the cause as obvious... e.g. maybe there are 3 - 4 different things that are needed to get a special badge, so if a project doesn't have the badge it isn't as easy to pin point why

**Reply by ianling** [reply]
I think some kind of "This package adheres to our guidelines for security best practices 👍" badge could at least be a good first step in the right direction for rewarding people who are doing it right and driving traffic to those packages. Anything is better than being presented with two options and having no quick way to distinguish between them besides Github stars.

**Reply by ianling** [reply]
I wonder if adding an option to proudly announce that you have enabled 2FA on your profile/on your package's profile would be a good middle ground. Not having this enabled doesn't necessarily imply that 2FA is disabled, just that you haven't chosen to broadcast that information to the world. In general, I think it could be a good way to give people who are doing everything right the option to put the spotlight on themselves, without inadvertently shining a light on people who are doing it wrong.

**Reply by lorsatti** [reply]
> We are actively discussing ways for us to improve 2FA adoption and overall security of the registry, but for the reasons you outlined above I do not believe that showing this status on a packages page is the correct approach.

@MylesBorins Can you please point us to these discussions? I see a need for it, if you look what is going on at faisalman/ua-parser-js#536. Thanks!

**Reply by ianling** [reply]
Two more from yesterday:
https://github.com/dominictarr/rc/issues/131
https://github.com/veged/coa/issues/99

coa has a single maintainer in npm: https://www.npmjs.com/package/coa

rc has multiple maintainers: https://www.npmjs.com/package/rc

**Reply by lorsatti** [reply]
> Two more from yesterday: [dominictarr/rc#131](https://github.com/dominictarr/rc/issues/131) [veged/coa#99](https://github.com/veged/coa/issues/99)
> 
> coa has a single maintainer in npm: https://www.npmjs.com/package/coa
> 
> rc has multiple maintainers: https://www.npmjs.com/package/rc

@ianling thanks, this is exactly what I mean, people are starting to ask for the second alternative you considered:

> Alternatively, starting on some arbitrary date, like 2022/01/01, all packages could be forced to have 2FA enabled in order to push changes.

So have a main discussion to strive for a possible, (definitive?) solution could be helpful.

**Reply by ianling** [reply]
https://github.blog/2021-11-15-githubs-commitment-to-npm-ecosystem-security/

tl;dr: they are going to make 2FA mandatory on certain popular packages in 2022. There were some security vulnerabilities in npm itself, but they are not related to the three instances of account takeovers mentioned above and have been fixed.

I think this is a step in the right direction, and I hope to see this expanded soon.

**Reply by ianling** [reply]
@MylesBorins this more recent article mentions plans to roll out enhanced login verification to all users on March 1st: https://github.blog/2022-02-01-top-100-npm-package-maintainers-require-2fa-additional-security/

Did that happen? I couldn't find any updates after this post.

**Reply by MylesBorins** [reply]
It did indeed happen and was documented via the GitHub change log

https://github.blog/changelog/2022-03-01-all-npm-accounts-are-now-enrolled-in-login-verification/

**Comment by MylesBorins** [comment]
I've unmarked my answer on this one as I'm rethinking our approach to see if there is a way we can introduce a feature like this without creating additional risk

**Comment by MylesBorins** [comment, marked as answer]
It has been quite some time and after much discussion it still seems like there is no way to share this information without making packages without 2FA at higher risk to attack. This is not something we will be doing
